compiler/vm/vm.nim

```diff
@@ -107,145 +181,236 @@ proc stackTraceImpl(
   c.config.handleReport(report, infoOrigin, action)
 
 
+# TODO: visit all usages of `stackTrace` and make sure that they abort the vm execution
 template stackTrace(
-    c: PCtx,
-    tos: PStackFrame,
+    c: TCtx,
+    tos: StackFrameIndex,
     pc: int,
     sem: ReportTypes,
     info: TLineInfo,
   ) =
   stackTraceImpl(c, tos, pc, info, instLoc())
   localReport(c.config, info, sem)
-  return
+  #return
 
 template stackTrace(
-    c: PCtx,
-    tos: PStackFrame,
+    c: TCtx,
+    tos: StackFrameIndex,
     pc: int,
     sem: ReportTypes,
   ) =
   stackTraceImpl(c, tos, pc, c.debug[pc], instLoc())
   localReport(c.config, c.debug[pc], sem)
-  return
+  #return
+
+func setNodeValue(dest: TypedHandle, node: PNode) =
+  assert dest.typ.kind == akPNode
+  GC_ref(node)
+  GC_unref(deref(dest).nodeVal)
+  deref(dest).nodeVal = node
+
 
-proc reportException(c: PCtx; tos: PStackFrame, raised: PNode) =
+
+template byteView(handle: TypedHandle): untyped =
+  handle.h.getMemoryRange(handle.typ.sizeInBytes)
+
+
+func copyToLocation(dest, src: TypedHandle, mm: var VmMemoryManager, reset: static[bool] = true) =
+  assert dest.typ == src.typ
+  copyToLocation(mm, dest.byteView(), src.h, src.typ, reset)
+
+
+func putIntoReg(reg: var TFullReg, v: sink VmObjectAtom, mm: var VmMemoryManager) =
+  case v.kind
+  of akInt:
+    reg = TFullReg(kind: rkInt, intVal: v.intVal)
+  of akFloat:
+    reg = TFullReg(kind: rkFloat, floatVal: v.floatVal)
+  of akPtr:
+    reg = TFullReg(kind: rkAddress, addrVal: makeHandle(v.ptrVal, v.typ.targetType.sizeInBytes), addrTyp: v.typ.targetType, isValid: true)
+  of akDiscriminator:
+    reg = TFullReg(kind: rkInt, intVal: readDiscriminant(v.atom, v.typ.sizeInBytes, v.typ.numBits))
+  of akPNode:
+    # TODO: ensure that only NimNodes go into registers (typedescs are akPNodes too)
```

Comment:
They are used during constant expression involving types (such as inside a `when`'s conditions). For example:
```
type A[T] = object
  when T is int:
    ...
```
The `T is int` expression is evaluated by the VM and `T` and `int` would be `typedesc`s here. For as to why this not done by `semfold`, I don't know. Maybe this can be changed, but for the scope of this PR I'd simply keep supporting it.

There also seem to be constant expressions yielding a `typedesc`, but I haven't looked into it further